### PR TITLE
WebRTC circumvention #588

### PR DESCRIPTION
**Pull request by atropnikov**
#588

### CODE REVIEW

File: Extension/lib/content-script/preload.js [hunk 1]
```diff
@@ -126,52 +126,84 @@
...
+
+        var scripts = [];
+
         // WebSockets are broken in old versions of chrome
         // and we don't need this hack in new version cause then websocket traffic is intercepted
         // https://github.com/AdguardTeam/AdguardBrowserExtension/issues/273
         // https://github.com/AdguardTeam/AdguardBrowserExtension/issues/572
-        var userAgent = navigator.userAgent.toLowerCase();
         var cIndex = userAgent.indexOf('chrome/');
         if (cIndex > 0) {
```
Comment (Mizzick) [inline Extension/lib/content-script/preload.js]: So it was except Chrome in specified version ranges but now it is only for Chrome? Do you miss other browsers?

Comment (atropnikov) [inline Extension/lib/content-script/preload.js]: Chrome, Edge, YaBrowser, Opera contain `Chrome/<version>` in the user agent string.

File: Extension/lib/content-script/preload.js [hunk 2]
```diff
@@ -126,52 +126,84 @@
...
             var version = userAgent.substring(cIndex + 7);
             var versionNumber = Number.parseInt(version.substring(0, version.indexOf('.')));
-            if (versionNumber < 47 || versionNumber > 57) {
-                return;
+            if (versionNumber >= 47 && versionNumber <= 57 &&
+                // This is chrome-specific feature for blocking WebSocket connections
+                // overrideWebSocket function is not defined in case of other browsers
```
Comment (ameshkov) [inline Extension/lib/content-script/preload.js]: It is not defined isn't it?

File: Extension/lib/content-script/preload.js [hunk 3]
```diff
@@ -126,52 +126,84 @@
...
 
         // Only for dynamically created frames and http/https documents.
         if (!isHtml() &&
             window.location.href !== "about:blank") {
             return;
         }
 
+        var userAgent = navigator.userAgent.toLowerCase();
+        if (userAgent.indexOf('firefox') >= 0) {
+            // Explicit check, we must not go further in case of Firefox
```
Comment (ameshkov) [inline Extension/lib/content-script/preload.js]: Don't we now need it to override WebRTC in Firefox?

File: Extension/lib/filter/rules/simple-regex.js [hunk 4]
```diff
@@ -32,7 +32,7 @@
 
             regexAnySymbol: ".*",
             regexSeparator: "([^ a-zA-Z0-9.%]|$)",
-            regexStartUrl: "^(http|https|ws|wss)://([a-z0-9-_.]+\\.)?",
+            regexStartUrl: "^(http://|https://|ws://|wss://|stun:|stuns:|turn:|turns:)([a-z0-9-_.]+\\.)?",
```
Comment (ameshkov) [inline Extension/lib/filter/rules/simple-regex.js]: Wow, that is a rather complicated regex.

First of all, we can change `stun:` to `stun://`, etc.

Then we'll be able to change this expression into something more readable:
`^.{2,5}://([a-z0-9-_.]+\\.)?`

File: Extension/lib/pages/log.js [hunk 5]
```diff
@@ -55,6 +55,20 @@ var StringUtils = {
 var UrlUtils = {
 
 	getDomainName: function (url) {
+        if (!url) {
+            return null;
+        }
+        if (url.indexOf('//') < 0) {
+            var index = url.indexOf(':');
```
Comment (ameshkov) [inline Extension/lib/pages/log.js]: We won't need it if we use `stun://` prefix

File: Extension/lib/utils/url.js [hunk 6]
```diff
@@ -40,50 +40,35 @@
...
             var firstIdx = url.indexOf("//");
-            if (firstIdx == -1) {
-                return null;
+            if (firstIdx === -1) {
+                /**
+                 * It's non hierarchical structured URL (e.g. stun: or turn:)
+                 * https://tools.ietf.org/html/rfc4395#section-2.2
+                 * https://tools.ietf.org/html/draft-nandakumar-rtcweb-stun-uri-08#appendix-B
+                 */
+                firstIdx = url.indexOf(":");
```
Comment (ameshkov) [inline Extension/lib/utils/url.js]: Again, we won't need it if we use `stun://`.

I realize it is against RFC, but anyway

File: Extension/lib/content-script/preload.js (hunk 3 above)
Comment (atropnikov) [inline Extension/lib/content-script/preload.js]: It was AMO reviewer's requirement.
See task and commit: https://github.com/AdguardTeam/AdguardBrowserExtension/issues/379

**Comment by ameshkov** [comment]
So, what with my comments? `stun:` thing?

**Comment by atropnikov** [comment]
We've decided not to replace the original `stun:` with `stun://`. `||` shouldn't cover WebRTC schemes and also if we do this, we can break rules like `stun:ads.com`.

**Comment by ameshkov** [comment]
Ok then, let's merge it.

**Comment by ameshkov** [comment]
@atropnikov please take a look at the latest changes in ABP, they've improved WebRTC wrapper a bit. Should we also?

**Comment by atropnikov** [comment]
@ameshkov Yes. These changes are in pull #670

**Pull request merged**